File: year2/mpp/java/MotorcycleContest/src/main/java/services/MotorcycleContestService.java
package services;

import domain.dtos.ParticipantInfo;
import domain.dtos.RaceInfo;
import domain.entities.Participant;
import domain.entities.Race;
import domain.entities.User;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import repositories.interfaces.ParticipantRepository;
import repositories.interfaces.RaceRepository;
import repositories.interfaces.TeamRepository;
import repositories.interfaces.UserRepository;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

@Component
public class MotorcycleContestService {
    private final UserRepository userRepository;
    private final RaceRepository raceRepository;
    private final TeamRepository teamRepository;
    private final ParticipantRepository participantRepository;

    @Autowired
    public MotorcycleContestService(UserRepository userRepository, RaceRepository raceRepository, TeamRepository teamRepository, ParticipantRepository participantRepository) {
        this.userRepository = userRepository;
        this.raceRepository = raceRepository;
        this.teamRepository = teamRepository;
        this.participantRepository = participantRepository;
    }

    public Optional<User> authenticate(String username, String password) {
        return userRepository.findByUsernameAndPassword(username, password);
    }

    public List<RaceInfo> getRacesBetween(Integer capacity1, Integer capacity2) {
        List<RaceInfo> races = new ArrayList<>();
        for (Race race : raceRepository.findByCapacityInterval(capacity1, capacity2)) {
            List<Participant> participants = participantRepository.findByRace(race.getId());
            races.add(new RaceInfo(race, participants.size()));
        }
        return races;
    }

    public RaceInfo getRaceByCapacity(Integer capacity) {
        Integer raceId = raceRepository.findIdByCapacity(capacity);
        Optional<Race> race = raceRepository.findOne(raceId);
        List<Participant> participants = participantRepository.findByRace(raceId);
        return new RaceInfo(race.get(), participants.size());
    }

    public List<RaceInfo> getRaceInfos() {
        List<RaceInfo> races = new ArrayList<>();
        for (Race race : raceRepository.findAll()) {
            List<Participant> participants = participantRepository.findByRace(race.getId());
            races.add(new RaceInfo(race, participants.size()));
        }
        return races;
    }

    public Iterable<Race> getRaces() {
        return raceRepository.findAll();
    }

    public List<ParticipantInfo> getRaceParticipantsByTeam(String teamName) {
        List<ParticipantInfo> participants = new ArrayList<>();
        Integer teamId = teamRepository.findIdByName(teamName);
        if (teamId == 0) {
            return participants;
        }
        for (Participant participant : participantRepository.findByTeam(teamId)) {
            Optional<Race> race = raceRepository.findOne(participant.getRaceId());
            race.ifPresent(value -> participants.add(new ParticipantInfo(value, participant)));
        }
        return participants;
    }

    public void registerParticipant(String name, String teamName, Integer capacity) {
        Integer teamId = teamRepository.findIdByName(teamName);
        Integer raceId = raceRepository.findIdByCapacity(capacity);
        participantRepository.save(new Participant(name, raceId, teamId));
    }
}

File: year2/mpp/c#/MotorcycleContest/MotorcycleContest/Services/MotorcycleContestService.cs
using System;
using System.Collections.Generic;
using MotorcycleContest.Domain.Dtos;
using MotorcycleContest.Domain.Entities;
using MotorcycleContest.Repositories.Interfaces;

namespace MotorcycleContest.Services
{
    public class MotorcycleContestService
    {
        private readonly IUserRepository _userRepository;
        private readonly IRaceRepository _raceRepository;
        private readonly ITeamRepository _teamRepository;
        private readonly IParticipantRepository _participantRepository;

        public MotorcycleContestService(IUserRepository userRepository, IRaceRepository raceRepository,
            ITeamRepository teamRepository, IParticipantRepository participantRepository)
        {
            this._userRepository = userRepository;
            this._raceRepository = raceRepository;
            this._teamRepository = teamRepository;
            this._participantRepository = participantRepository;
        }

        public User Authenticate(string username, string password)
        {
            return _userRepository.FindByUsernameAndPassword(username, password);
        }

        public List<RaceInfo> GetRacesBetween(int capacity1, int capacity2)
        {
            List<RaceInfo> races = new List<RaceInfo>();
            foreach (Race race in _raceRepository.FindByCapacityInterval(capacity1, capacity2)) {
                List<Participant> participants = _participantRepository.FindByRace(race.ID);
                races.Add(new RaceInfo(race, participants.Count));
            }
            return races;
        }

        public RaceInfo GetRaceByCapacity(int capacity)
        {
            int raceId = _raceRepository.FindIdByCapacity(capacity);
            Race race = _raceRepository.FindOne(raceId);
            List<Participant> participants = _participantRepository.FindByRace(race.ID);
            return new RaceInfo(race, participants.Count);
        }

        public IEnumerable<Race> GetRaces() {
            return _raceRepository.FindAll();
        }
        
        public List<ParticipantInfo> GetRaceParticipantsByTeam(string teamName) {
            List<ParticipantInfo> participants = new List<ParticipantInfo>();
            int teamId = _teamRepository.FindIdByName(teamName);
            if (teamId == 0) {
                return participants;
            }
            foreach (Participant participant in _participantRepository.FindByTeam(teamId)) {
                Race race = _raceRepository.FindOne(participant.RaceId);
                if (race != null)
                {
                    participants.Add(new ParticipantInfo(race, participant));
                }
            }
            return participants;
        }

        public void RegisterParticipant(string name, string teamName, int capacity) {
            int teamId = _teamRepository.FindIdByName(teamName);
            int raceId = _raceRepository.FindIdByCapacity(capacity);
            Participant participant = _participantRepository.Save(new Participant(name, raceId, teamId));
        }
    }
}